### PR TITLE
BUG: Position is not offset

### PR DESCRIPTION
**Pull request by aylward**
Resolves issue https://github.com/Kitware/MetaIO/issues/127

Previously MetaIO considered origin and offset to have the same
meaning and transformMatrix and direction to have the same meaning.
However, in SpatialObjects, which are read and written as metaobjects,
position and offset have different meanings and so do transformMatrix
and direction.  So, prior to this commit, saving a scene graph of
spatialobjects could result in the loss of critical info.

ElementOrigin is introduced as the location of an object in its inherent
/ isolated / original space.  In particular, for images, it is the
"Image Position (patient)" tag which is the itk Origin() definition.

ElementDirection is introduced as the direction of acquisition for
the image. It follows the ITK definition of Direction().

Offset and transformMatrix (and centerofrotation) specify the
object-to-parent-object transform.  That is how an object (and all of
its children objects, aka. its scene subgraph) are transformed to map
into the parent object space.

Backward compatibility is preserved.   MetaIO defaults to using the
original API and the original file format.  If you wish to use the new
meanings (API) and write in the new file format, set the
APIVersion and FileFormatVersion variables to 2 (instead of 1 - default).

### CODE REVIEW

File: src/metaArrow.cxx
Comment (dzenanz) [inline src/metaArrow.cxx:53]: Why is the arrow the only one to receive special updates (in addition to image)?

Comment (aylward) [reply src/metaArrow.cxx:53]: They are the only ones which were inconsistent.   Tubes, blobs, boxes, contours, etc already existed in their own coordinate systems.  Arrow only had a length and a direction prior to this.  The object-to-parent transform had to be used to move the arrow to a position.  It really made handling annotations inconsistent.   This change is conceptually very helpful - particularly for maskImages and images.  I swear this all worked correctly years ago when Julien and I wrote spatial object viewers, but I haven't tried to go back and look at where things went wonky between metaIO and spatialObjects.

Comment (aylward) [reply src/metaArrow.cxx:53]: And please don't merge yet - note I changed name to WIP.

I found a memory issue when handling user-defined fields.   It was very brittle - it worked if you called the exact right functions in the exact right order.  I've updated to make it much more logical and robust.   Will include another commit in this pull request.  Also, will rebase and squash all before merging.

Thanks!

Comment (dzenanz) [reply src/metaArrow.cxx:53]: I believe you can convert the PR to `Draft` status.

**Comment by aylward** [comment]
@dzenanz - feel free to do a final review.  All tests pass when built stand-alone or as part of ITK.  

I will squash and merge if you approve or if you pass on reviewing after a few days. :)

I'll then do a PR to update MetaIO in ITK and associated changes to itkSpatialObjects.   So happy to have this working again.

**Review by dzenanz — APPROVED** [review]
Looks good on a glance.

**Pull request merged**
